Add support for non-Tuple iterators

dennisYatunin opened the pull request:
This PR adds an interface to `UnrolledUtilities` that allows it to handle non-`Tuple` iterators as both inputs and outputs of unrolled functions.

## Content
- Adds what is essentially a simplified version of the `Base.AbstractBroadcasted` interface for static iterators (a detailed summary is located at the top of `src/unrollable_interface.jl`).
- Adds support for lazy iterators from `Base`, such as those generated by the `enumerate`, `zip`, and `Iterators.map` functions. Also removes the `unrolled_zip` and `unrolled_enumerate` functions, since they are no longer necessary. These functions were not used outside of this repository, so this shouldn't be a breaking change.
- Adds support for `NamedTuple`s and iterators from `StaticArrays`.
- Implements and adds support for static lazy iterators, which store unevaluated computations that can be inlined elsewhere, and static bitwise iterators, which reduce the register space needed by a collection of `Bool`s by a factor of 8 in comparison to `NTuple{N, Bool}`.
- Adds some tests to demonstrate when it is beneficial to use static bitwise iterators. As is shown in the auto-generated comparison table, passing an iterator larger than 32 bytes in size through at least two levels of reduction/accumulation results in allocations, even when everything is unrolled. So, replacing a `Tuple` with a static bitwise iterator can allow us to use 256 `Bool`s instead of 32 without triggering allocations, which will be useful for optimizing the non-orographic gravity wave tendency in `ClimaAtmos`. (cc @Xiaoyang-Xie)
- Adds an unrolled version of `Base.accumulate`, along with an optimized version for static bitwise iterators. This could also be helpful for the gravity wave tendency.
- Improves the auto-generated comparison table by measuring compilation time and memory usage more accurately, splitting memory usage measurements into GC (garbage collector) and RSS (resident set size) values, adding measurements of runtime, and summarizing compiler optimizations more clearly.

dennisYatunin committed:
Add support for non-Tuple iterators, improve docs and tests

charleskawczynski commented on ext/UnrolledUtilitiesStaticArraysExt.jl:
```diff
@@ -0,0 +1,10 @@
+
+UnrolledUtilities.length_from_type(::Type{<:SVector{N}}) where {N} = N
+UnrolledUtilities.target_output_type(::SVector) = SVector
+UnrolledUtilities.output_constructor(::Type{SVector}) = SVector
```
```suggestion
UnrolledUtilities.output_constructor(::Type{<:SVector}) = SVector
```
Should this subtyping?

dennisYatunin replied:
The input to `output_constructor` will always be a type returned by `target_output_type`. The type in this case is just `SVector`, so we don't need to subtype here. Same for `Tuple` and `LazySequence`, though we need to subtype for `BitSequence` to propagate the unsigned int type.

charleskawczynski commented on src/BitSequence.jl:
```diff
@@ -0,0 +1,148 @@
+target_output_type(::BitSequence{<:Any, U}) where {U} = BitSequence{<:Any, U}
+
+output_promote_rule(::Type{B}, ::Type{O}) where {B <: BitSequence, O} = O
+output_promote_rule(::Type{B}, ::Type{Tuple}) where {B <: BitSequence} = Tuple
```
Where/how is this used?

dennisYatunin replied:
The main motivation for `output_promote_rule` is to allow mixing of `BitSequence`s with other iterator types without losing efficiency. For example, when a `BitSequence` and a `LazySequence`/`LazyMap` are passed to `unrolled_map` or `unrolled_accumulate`, `output_promote_rule` makes the output a `BitSequence` and ensures that the specialized method of `unrolled_map` or `unrolled_accumulate` for `BitSequence`s is used (as long as the output elements are all `Bool`s).

The two promotion rules highlighted here ensure that mixing a low-storage `BitSequence` with a high-storage `Tuple` or some other type generates the high-storage type as an output, since there is no benefit to generating low-storage iterators when high-storage iterators are already using up register space.

charleskawczynski commented on src/BitSequence.jl:
```diff
@@ -0,0 +1,148 @@
+    return Bool(int >> bit_offset & one(int))
+end
+
+@inline function Base.setindex(itr::BitSequence, bit::Bool, n::Integer)
```
I think we'd want to use `Base.@propagate_inbounds` here

charleskawczynski commented on src/BitSequence.jl:
```diff
@@ -0,0 +1,148 @@
+    return (int_offset + 1, bit_offset)
+end
+
+@inline function Base.getindex(itr::BitSequence, n::Integer)
```
I think we'd want to use `Base.@propagate_inbounds` here

charleskawczynski reviewed (approved):
I discussed the design with @dennisYatunin, and it looks good. The generated table is really nice for seeing how things perform, too.

It'll be nice to use this new version in ClimaCore

charleskawczynski commented on test/test_and_analyze.jl:
```diff
@@ -118,7 +174,35 @@ function code_instance(f, args...)
-macro test_unrolled(args_expr, unrolled_expr, reference_expr, contents_info_str)
+macro benchmark(expression)
+    return quote
+        Base.Experimental.@force_compile
```
I'm wondering if we should use this-- it puts the test suite at risk for breaking without notice on version updates.

dennisYatunin replied:
Looks like we didn't need the `@force_compile` here; I just pushed an update that removes it. As for the use of `@force_compile` for the compilation time measurements, we need it in order to get accurate results, and it won't break when we switch to 1.11 because the same code is still being used in [base/timing.jl](https://github.com/JuliaLang/julia/blob/master/base/timing.jl). Also, when we switch to 1.11, we'll be able to use the publicly exported quantity `(@timed <expr>).compile_time`. I can open an issue to remember to do that after this is merged in.

charleskawczynski reviewed (changes requested):
I’ve thought about this more, let’s remove the experimental piece.

The simplest way to force compilation is to simply call the method with a type signature. Let’s do that instead.

dennisYatunin commented:
@charleskawczynski It looks like the `@force_compile` annotation is only necessary for accurate timing results when running Julia interactively. Since the test suite runs each measurement non-interactively in a separate process, we can remove the annotation without affecting the results.

charleskawczynski commented on src/StaticOneTo.jl:
```diff
@@ -0,0 +1,18 @@
+for `unrolled_take`, but no other unrolled functions can use `StaticOneTo`s as
+output types.
+"""
+struct StaticOneTo{N} <: StaticSequence{N} end
```
Is there a reason why we're not just using `StaticArrays.SOneTo`? (it's an exported/documented type)

charleskawczynski replied:
Ah, this subtypes StaticSequence

dennisYatunin replied:
`StaticOneTo` is used to define `StaticBitVector`. If we were to use `StaticArrays.SOneTo` instead, we would need to make `StaticArrays` a dependency of the package (right now it is just an extension). The definition of `StaticOneTo` is also just 5 lines, which seems like an acceptable tradeoff for avoiding a dependency.

charleskawczynski commented:
Will this close #9?

dennisYatunin commented:
@charleskawczynski Yes, the inference failures demonstrated in the ClimaCore PR linked in #9 were caused by UnrolledUtilities being unable to handle `NamedTuple`s. This PR adds full support for `NamedTuple`s.

Also, I've added some benchmarks to compare generative unrolling (what this package was previously doing) with recursive unrolling (what lots of functions in ClimaCore are doing). It turns out that recursive unrolling is faster to compile for small iterators, though it is much slower to compile and generates less optimal code for large iterators. I've modified the interface so that we can switch from generative to recursive unrolling for small iterators (the cutoff is currently set to 16). Since nearly all of the unrolling we do in ClimaCore is over small iterators, this should allow us to use UnrolledUtilities throughout ClimaCore without any increase in compilation time.

dennisYatunin commented:
@Sbozzolo I have added a bunch of documentation to this PR, with a focus on guiding users through the all of the different problems addressed by `UnrolledUtilities`. Could you please take a look and let me know whether it is sufficiently clear? In particular, is it enough for UnrolledUtilities to be used in ClimaCore and ClimaAtmos? If so, we can move ahead with CliMA/ClimaAtmos.jl#3267.

charleskawczynski commented:
> @charleskawczynski Yes, the inference failures demonstrated in the ClimaCore PR linked in https://github.com/CliMA/UnrolledUtilities.jl/issues/9 were caused by UnrolledUtilities being unable to handle NamedTuples. This PR adds full support for NamedTuples. Also, I've added some benchmarks to compare generative unrolling (what this package was previously doing) with recursive unrolling (what lots of functions in ClimaCore are doing). It turns out that recursive unrolling is faster to compile for small iterators, though it is much slower to compile and generates less optimal code for large iterators. I've modified the interface so that we can switch from generative to recursive unrolling for small iterators (the cutoff is currently set to 16). Since nearly all of the unrolling we do in ClimaCore is over small iterators, this should allow us to use UnrolledUtilities throughout ClimaCore without any increase in compilation time.

Good work, and I think that makes sense. Thanks for taking this on.

dennisYatunin commented:
After some offline discussion, it looks like everyone is in agreement that the docs are now usable. The are still a few unit tests I'd like to add, and also a lingering issue regarding `unrolled_reduce(f, ::StaticOneTo)` vs. `unrolled_reduce(f, ::Val)` that could be affecting compilation times in ClimaAtmos, but I will address these in a future PR.
